src/checkpoint.hh

```diff
@@ -0,0 +1,1246 @@
+/* {{{
+
+    This file is part of gapc (GAPC - Grammars, Algebras, Products - Compiler;
+      a system to compile algebraic dynamic programming programs)
+
+    Copyright (C) 2008-2011  Georg Sauthoff
+         email: gsauthof@techfak.uni-bielefeld.de or gsauthof@sdf.lonestar.org
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+    * Author: fymue
+}}} */
+
+#ifndef SRC_CHECKPOINT_HH_
+#define SRC_CHECKPOINT_HH_
+
+#include <string>
+#include <array>
+#include <vector>
+#include <list>
+#include <utility>
+#include <sstream>
+#include "printer.hh"
+#include "symbol.hh"
+#include "statement/table_decl.hh"
+#include "type/base.hh"
+
+typedef hashtable<std::string, Symbol::NT*> nt_tables;
+
+namespace Printer {
+class Checkpoint : public Base {
+ private:
+  // true if table types are wrapped in ListRef
+  bool list_ref;
+
+  std::vector<std::string> string_type_accessors, block_type_accessors,
+                           subseq_type_accessors;
+
+  // currently supported/serializable datatypes
+  const std::array<Type::Type, 11>
+  SUPPORTED_TYPES = {Type::Type::VOID, Type::Type::INTEGER,
+                     Type::Type::INT, Type::Type::FLOAT,
+                     Type::Type::SIZE, Type::Type::SINGLE,
+                     Type::Type::BIGINT, Type::Type::STRING,
+                     Type::Type::SHAPE, Type::Type::SUBSEQ,
+                     Type::Type::EXTERNAL};
+
+const std::array<std::string, 12>
+SUPPORTED_EXTERNAL_TYPES = {"Rope", "answer_pknot_mfe", "pktype",
```

Comment:
is Rope really considered extern? Isn't it part of rtlib? As well as shape_t?

Reply:
Yes, `Rope` is considered extern. `Shape` isn't, but in the `pKiss` programs of the fold-grammars repo a different alphabet is used for the `Shape`/`Fiber` type and the `shape_t` typedef is overwritten, which makes it extern in that case. If the regular gapc `Shape` is used it will be enumerated as `Type::SHAPE`.